python/lsst/summit/utils/tmaUtils.py

```diff
@@ -0,0 +1,1588 @@
+            return None
+
+        # check case 2)
+        if time > events[-1].end:
+            self.log.warning(f'{logStart} and is after the last event of the day')
+            return None
+
+        # we are now either in an event, or between events. Walk through the
+        # events, and if the end of the event is after the specified time, then
+        # we're either in it or past it, so check if we're in.
+        for eventNum, event in enumerate(events):
+            if event.end > time:  # case 3) we are now into or past the right event
+                # the event end encloses the time, so note the > and not >=,
+                # this must be strictly greater, we check the overlap case
+                # later
+                if time >= event.begin:  # we're fully inside the event, so return it.
+                    # 3a) before returning, check if we're exactly at the start
+                    # of the event, and if so, log it. Then 3b) also check if
+                    # we're at the exact end of the previous event, and if so,
+                    # log that too.
+                    if time == event.begin:
+                        self.log.info(f"{logStart} and is exactly at the start of event"
+                                      f" {eventNum}")
+                        if eventNum == 0:  # I think this is actually impossible, but check anyway
+                            return event  # can't check the previous event so return here
+                        previousEvent = events[eventNum - 1]
+                        if previousEvent.end == time:
+                            self.log.info("Previous event is contiguous, so this time is also at the exact"
+                                          f" end of {eventNum - 1}")
+                    return event
+                else:  # case 4)
+                    # the event end is past the time, but it's not inside the
+                    # event, so we're between events. Log which we're between
+                    # and return None
+                    previousEvent = events[eventNum - 1]
+                    timeAfterPrev = (time - previousEvent.end).to_datetime()
+                    naturalTimeAfterPrev = humanize.naturaldelta(timeAfterPrev, minimum_unit='MICROSECONDS')
+                    timeBeforeCurrent = (event.begin - time).to_datetime()
+                    naturalTimeBeforeCurrent = humanize.naturaldelta(timeBeforeCurrent,
+                                                                     minimum_unit='MICROSECONDS')
+                    self.log.info(f"{logStart} and lies"
+                                  f" {naturalTimeAfterPrev} after the end of event {previousEvent.seqNum}"
+                                  f" and {naturalTimeBeforeCurrent} before the start of event {event.seqNum}."
+                                  )
+                    return None
+
+        raise RuntimeError('This should never happen')
```

Comment:
You know if you write `'This should never happen'` some frustrated programmer will be trying to debug it in eight years, so consider writing a slightly more descriptive error message. I *think* you will actually hit this if `time = events[-1].end`

Reply:
Ha, you're absolutely right! Added a failing test for this, and will fix that now. Will try to improve the error message a bit too 🙂